test/python/pulse/test_builder.py

```diff
@@ -0,0 +1,969 @@
+        # d1
+        reference.insert(3, instructions.Delay(5, d1), mutate=True)
+
+        self.assertEqual(schedule, reference)
+
+    def test_align_left(self):
+        """Test the left alignment context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+        d2 = pulse.DriveChannel(2)
+
+        with pulse.build() as schedule:
+            with pulse.align_left():
+                pulse.delay(11, d2)
+                pulse.delay(3, d0)
+                with pulse.align_left():
+                    pulse.delay(5, d1)
+                    pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference += instructions.Delay(3, d0)
+        reference += instructions.Delay(7, d0)
+        # d1
+        reference = reference.insert(3, instructions.Delay(5, d1))
+        # d2
+        reference += instructions.Delay(11, d2)
+        self.assertEqual(schedule, reference)
+
+    def test_align_right(self):
+        """Test the right alignment context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+        d2 = pulse.DriveChannel(2)
+
+        with pulse.build() as schedule:
+            with pulse.align_right():
+                pulse.delay(11, d2)
+                pulse.delay(3, d0)
+                with pulse.align_right():
+                    pulse.delay(5, d1)
+                    pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference = reference.insert(1, instructions.Delay(3, d0))
+        reference = reference.insert(4, instructions.Delay(7, d0))
+        # d1
+        reference = reference.insert(6, instructions.Delay(5, d1))
+        # d2
+        reference += instructions.Delay(11, d2)
+
+        self.assertEqual(schedule, reference)
+
+    def test_group(self):
+        """Test the grouping context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+
+        with pulse.build() as schedule:
+            pulse.delay(3, d0)
+            with pulse.group():
+                pulse.delay(5, d1)
+                pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference += instructions.Delay(3, d0)
+        reference += instructions.Delay(7, d0)
+        # d1
+        reference = reference.insert(3, instructions.Delay(5, d1))
+        self.assertEqual(schedule, reference)
+
+    def test_inline(self):
+        """Test the inlining context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+
+        with pulse.build() as schedule:
+            pulse.delay(3, d0)
+            with pulse.inline():
+                pulse.delay(5, d1)
+                pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference += instructions.Delay(3, d0)
+        reference += instructions.Delay(7, d0)
+        # d1
+        reference += instructions.Delay(5, d1)
```

Comment:
is there a test which would show `inline` behaving differently than not having it at all? What does it do?

Reply:
This flattens the instructions below into the calling Schedule. Is useful for example if you want to write a macro that just produces instructions that inherit the alignment of the parent context. 

Added better test in 3da37a4b5. Here you can see the alignment is ignored due to the inlining.